### PR TITLE
Fix: Send only blobs relevant to operators

### PR DESCRIPTION
**Pull request by ian-shim**
## Why are these changes needed?
For a given encoded blob, it keeps track of blob messages for given operator. If an operator is not part of a quorum and if a blob is dispersed to a quorum the operator is not part of, the encoded blob should not include the operator's blob message, but the blob header should be the same and sent to the operators even if the operator is not part of the quorum the blob is getting dispersed. 
This PR makes `BlobHeader` field in `EncodedBlob` common while having assigning different bundles for different operators. Also, it sends the blob headers to all operators (makes all operators store the blob headers). This is to allow operators to properly validate the batch root for all blob headers. If an operator isn't part of the quorum the blob is dispersed to, it will receive an empty bundle. 

## Checks

- [x] I've made sure the lint is passing in this PR.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: disperser/batcher/grpc/dispatcher.go [hunk 1]
```diff
@@ -153,15 +155,18 @@ func getBlobMessage(blob *core.BlobMessage) (*node.Blob, error) {
...
 
-	for i, header := range blob.BlobHeader.QuorumInfos {
-		quorumHeaders[i] = &node.BlobQuorumInfo{
-			QuorumId:           uint32(header.QuorumID),
-			AdversaryThreshold: uint32(header.AdversaryThreshold),
-			ChunkLength:        uint32(header.ChunkLength),
-			QuorumThreshold:    uint32(header.QuorumThreshold),
-			Ratelimit:          header.QuorumRate,
+	for _, header := range blob.BlobHeader.QuorumInfos {
+		if _, ok := blob.Bundles[header.QuorumID]; ok {
```
Comment (mooselumph) [inline disperser/batcher/grpc/dispatcher.go:161]: Shoudn't this always be ok?

**Review by mooselumph — APPROVED** [review]
It makes sense to me. Good catch. Just one question.

File: disperser/batcher/grpc/dispatcher.go (hunk 1 above)
Comment (ian-shim) [reply disperser/batcher/grpc/dispatcher.go:161]: The way it's implemented today, `blob.BlobHeader` is the same for all blobs sent to all operators. So, when an operator is only registered in quorum 1 and blob is dispersed to quorums 0 and 1, `blob.BlobHeader.QuorumInfos` will contain both quorums. 
However, `blob.Bundles` is operator specific, so for operator only in quorum 1, `blob.Bundles[0]` does not exist.

Comment (mooselumph) [reply disperser/batcher/grpc/dispatcher.go:161]: Wait, the DA nodes should be verifying the merkle root in the batch header, right? If they aren't, it would be a security hole. 

In order for them to verify the root, they need all the leaves, which means all of the blobHeaders.

File: core/data.go [hunk 2]
```diff
@@ -138,7 +138,10 @@ type BatchHeader struct {
 }
 
 // EncodedBlob contains the messages to be sent to a group of DA nodes corresponding to a single blob
-type EncodedBlob = map[OperatorID]*BlobMessage
+type EncodedBlob struct {
```
Comment (ian-shim) [inline core/data.go:141]: makes `BlobHeader` common across all operators

File: disperser/batcher/grpc/dispatcher.go [hunk 3]
```diff
@@ -153,28 +156,35 @@ func getBlobMessage(blob *core.BlobMessage) (*node.Blob, error) {
...
 	// the ordering of quorums in bundles must be same as in quorumHeaders
 	for i, quorumHeader := range quorumHeaders {
 		quorum := quorumHeader.QuorumId
-		bundles[i] = &node.Bundle{
-			Chunks: data[quorum],
+		if _, ok := blob.Bundles[uint8(quorum)]; ok {
+			bundles[i] = &node.Bundle{
+				Chunks: data[quorum],
+			}
+		} else {
```
Comment (ian-shim) [inline disperser/batcher/grpc/dispatcher.go:183]: Makes sure `len(quorumHeaders) == len(bundles)`
Bundles will be empty for quorums the operator is not part of

File: disperser/batcher/grpc/dispatcher.go [hunk 4]
```diff
@@ -153,28 +156,35 @@ func getBlobMessage(blob *core.BlobMessage) (*node.Blob, error) {
 		lengthProofData.YA1 = blob.BlobHeader.LengthProof.Y.A1.Marshal()
 	}
 
-	quorumHeaders := make([]*node.BlobQuorumInfo, len(blob.BlobHeader.QuorumInfos))
+	quorumHeaders := make([]*node.BlobQuorumInfo, 0)
```
Comment (mooselumph) [inline disperser/batcher/grpc/dispatcher.go:159]: Would it better just to revert the changes here?

**Comment by mooselumph** [comment]
looks good!

File: disperser/batcher/grpc/dispatcher.go (hunk 4 above)
Comment (ian-shim) [reply disperser/batcher/grpc/dispatcher.go:159]: Sure!

File: disperser/batcher/grpc/dispatcher.go [hunk 5]
```diff
@@ -34,23 +34,27 @@ func NewDispatcher(cfg *Config, logger common.Logger) *dispatcher {
...
+			blobMessages := make([]*core.BlobMessage, 0)
+			for _, blob := range blobs {
+				blobMessages = append(blobMessages, &core.BlobMessage{
+					BlobHeader: blob.BlobHeader,
+					// Bundles will be empty if the operator is not in the quorums blob is dispersed on
+					Bundles: blob.BundlesByOperator[id],
+				})
 			}
-			sig, err := c.sendChunks(ctx, blobMessages, header, &op)
+			sig, err := c.sendChunks(ctx, blobMessages, batchHeader, &op)
```
Comment (jianoaix) [inline disperser/batcher/grpc/dispatcher.go:57]: Why disperse to an operator if it's not in any quorum of a batch?

Comment (ian-shim) [reply disperser/batcher/grpc/dispatcher.go:57]: It should only disperse the blob headers to the operators not in the quorums. An operator needs all the blob headers (even the ones they don't participate on) to verify the batch root

Comment (jianoaix) [reply disperser/batcher/grpc/dispatcher.go:57]: Why do they need to verify the batch root? If the batch has quorum 1,2 (aggregating from blobs inside), and an operator is only in quorum 3, why it has to verify/sign a batch that it has no responsibility?
